#7 : Settings page

leo7418 opened the pull request:
https://github.com/FileShare-Project/file-share/issues/7#issue-2054091094

leo7418 committed:
- Add some assets
- New class Debug which handle debugging a widget and its nested children
- New components class:
    - Button: handle basics of button with aligment, icon and style
    - Foldout: handle collapsible container controlled by a button
    - List: handle vertical layout with space between items and separator of custom height
    - ListMenu: handle menu with titles, foldout items and subitems
- New MVC Settings:
    - Control settings of the application rendered in a ListMenu
- On Application class:
    - Setup the Settings MVC

leo7418 committed:
- On Button component:
    - Can now set a custom image size instead of a ratio
    - Can now set a distance between text and image
    - Fix right aligment
- On List component:
    - Store now signal ids of onSizeChange for each item
    - Disconnect only the signal associated to the removed widget instead of all its signals

leo7418 committed:
- On Button component:
    - Bug fixed: offset not used when there is no image
- On Foldout component:
    - Emit now signals onClick, onOpen and onClose
    - Can now set the image size
    - Can now define if the foldout is foldable or not
    - Bug fixed: updateContentHeight doesn't update the component size when already opened
    - Bug fixed: use setHeight when there is no animation instead of setSize. Otherwise the width is set to 0 (bug of tgui)
- On List component:
    - When removing a widget, recalcul the items position only from the removed widget
- On ListMenu component:
    - Set to foldout all options
    - Remove option textColor as it must be controlled by the button style
- On DeviceList View:
    - Use now the menu functions instead of adding custom widgets
    - Set foldable to false for menu items
    - Add some padding
- On Settings View:
    - Set defaultFolded to true

leo7418 committed:
- On CMakeLists:
    - Add sqlite_orm
- Add account and settings_device assets
- On Application:
    - Remove buildSiderButton and buildContent
    - Have now only one sider and a button for switching between deviceList and settings
- On Button component:
    - Can now set the size of a button with an enum
- On Foldout:
    - Can set the buttonSize
- On ListMenu:
    - Remove textSize and iconSize options as its now managed by the button component
    - Set buttonSize to Medium for items
    - Set buttonSize to Small for sub-items
- On DeviceList View and Settings View:
    - Inherit now from ListMenu
- On Settings View:
    - Add signal onSettingsClicked with the associated widget container for this settings as emit parameter

leo7418 committed:
- New component InputFileDialog
- On List component:
    - Set autoLayout to manual and autoLayoutUpdateEnabled to false to each added items
    - Use setHeight instead of setSize
- On Settings view:
    - Add settings for each section

leo7418 committed:
- On Button component:
    - Can now set the button in an active state
    - Can automatically manage active state
- On Foldout component:
    - Can now retrieve widgets of the content section
    - Can now enable the active state of the button
    - Can now retrieve if the foldout is open
    - Separate setFoldable in two functions for disabling the fold image
    - Bug fixed: finishAllAnimations before starting another one
- On ListMenu:
    - Can now set an active mode
    - New signals onMenuClicked, onMenuActive, onMenuInactive and onSelectionChange
- On Settings view:
    - Separate device settings section in 3 distincts section and sub-menus
    - Rename signal onSettingsClicked to onMenuChange
    - Handle the ListMenu onSelectionChange signal for emitting the onMenuChange signal
    - Can now retrieve the current selected menu
- On Application:
    - Bug fixed: remove all widgets when switching between device list and settings
    - Bug fixed: add to the main panel the current menu content

f-michaut commented on include/Components/Button.hpp:
```diff
@@ -0,0 +1,108 @@
+			typedef std::shared_ptr<const Button> ConstPtr;
+
+			static Button::Ptr create() { return std::make_shared<Button>(); }
+			static Button::Ptr copy(Button::ConstPtr widget) { return widget ? std::static_pointer_cast<Button>(widget->clone()) : nullptr; }
```
Indentation... We need to configure a `.editorconfig` for the project I think

f-michaut commented on include/Components/Button.hpp:
```diff
@@ -0,0 +1,108 @@
+			static Button::Ptr copy(Button::ConstPtr widget) { return widget ? std::static_pointer_cast<Button>(widget->clone()) : nullptr; }
+
+            void setImageSize(float size) { this->imageSize = {size, size}; this->setImageScaling(-1); }
+            void setImageSize(const tgui::Vector2f &size) { this->imageSize = size; this->setImageScaling(-1); }
```
```suggestion
            void setImageSize(float size) { this->setImageSize({size, size}); }
            void setImageSize(const tgui::Vector2f &size) { this->imageSize = size; this->setImageScaling(-1); }
```
Use overloaded functions to avoid repeating logic (currently repeating `setImageScaling`)

f-michaut commented on include/Components/Button.hpp:
```diff
@@ -0,0 +1,108 @@
+            bool isActive() const { return this->canBeActive() ? active : false; }
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<Button>(*this); }
```
intendation

f-michaut commented on include/Components/Button.hpp:
```diff
@@ -0,0 +1,108 @@
+            void updateStyle();
+            void updateActiveStyle();
+
+            std::shared_ptr<tgui::priv::dev::ImageComponent> imageComponent;
```
I dont think we are supposed to use what is inside of `tgui::priv::dev`, the extra namespaces are here to signify this is internal classes and are likely to break when changing library versions.

Is there something else from TGui we could use ?

leo7418 replied:
Unfortunately, it's the only way to control the button image.
I had talked about it on Discord with the creator of TGui, and he told me it was possible to use it if I knew what I was doing. I looked at the source code carefully, so it shouldn't be a problem.

Otherwise, I would have to completely re-implement the button class, including the different button states and the renderer, which would be a lot more work.

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+			typedef std::shared_ptr<const Foldout> ConstPtr;
+
+			static Foldout::Ptr create() { return std::make_shared<Foldout>(); }
+			static Foldout::Ptr copy(Foldout::ConstPtr widget) { return widget ? std::static_pointer_cast<Foldout>(widget->clone()) : nullptr; }
```
indentation

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<Foldout>(*this); }
+
```
indentation

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+            bool removeFromContent(const Widget::Ptr &widget);
+            const std::vector<tgui::Widget::Ptr> &getContentWidgets() const { return this->content->getWidgets(); }
+
+            void setButtonSize(Components::Button::Size size) { this->button->setSize(size); }
+            Components::Button::Size getButtonSize() const { return this->button->getSize(); }
+
+            void buttonCanBeActive(bool active) { this->button->canBeActive(active, false); }
+            bool buttonCanBeActive() const { return this->button->canBeActive(); }
+
+            void setText(const tgui::String &text) { this->button->setText(text); }
+            const tgui::String &getText() const { return this->button->getText(); }
+
+            void setImage(const tgui::String &image) { this->button->setImage(image); }
+            const tgui::Texture &getImage() const { return this->button->getImage(); }
+
+            void setImageSize(float size) { this->button->setImageSize(size); }
+            void setImageSize(const tgui::Vector2f &size) { this->button->setImageSize(size); }
+            tgui::Vector2f getImageSize() const { return this->button->getImageSize(); }
```
Instead of all theses functions, how about : 
```c++
  Button &getButton();
  const Button &getButton() const;
```
Calling code can do `foldout.getButton().setText("ABC");` instead of `foldout.setButtonText("ABC");` which is almost the same thing, but greatly reduces the amount of functions we have to write.

Furthermore, if TGui adds more functions to buttons, or we need to access a function of button we haven't exposed yet, we will have to add it manually. It increase maintenance by a lot.

leo7418 replied:
It was to create an abstraction, to separate the fact that the Foldout class uses the Button class.
And there won't be any more functions normally, I didn't include everything that was present in the Button class but only what can be configured for the Foldout class.

I assumed that if a foldout is created, we only have access to these functions in order not to modify the behavior of the button in the foldout, only the foldout can be manipulated this way.

f-michaut replied:
I dont think we need the abstraction, it adds boilerplate code and adds arbitrary restrictions if we want to use more features of the button (or other incapsulated class) later on. 

The only advantage is if we want to replace the `Components::Button` by another type of button, we would only have to update the boilerplate functions - but we are already using our custom button type, so I don't really see that happening

f-michaut replied:
This is different tho, the methods are just wrappers, so I really don't see the point

leo7418 replied:
They are wrappers, indeed, but they are important to properly separate the logic of the Button and the Foldout.
Moreover, if I leave full access to all the Button functions, nothing will prevent me from calling a Button function that would go against the Foldout, which would cause unexpected behavior and potentially complicated issues to resolve.

I see it more as a security measure rather than a simple wrapper.
Additionally, they are inline by default (if I recall correctly from the compilation behavior? And if not, I can add the `inline` keyword), so in terms of execution, it shouldn't change much.

If one day I really have too many functions to add concerning the Button, I can review this part of the code then. But currently, just for code safety, I would prefer to leave it as it is

f-michaut replied:
I dont think we need the security : we can be responsible and not use any function that would break the behavior of the foldout

Yeah they are inline since definition is in hpp, my concern wasn't performance, but maintainability.

We can keep them for now

f-michaut replied:
I dont think we need the security : we can be responsible and not use any function that would break the behavior of the foldout

Yeah they are inline since definition is in hpp, my concern wasn't performance, but maintainability.

We can keep them for now

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+            tgui::Signal onOpen = { "Opened" };
+            tgui::Signal onClose = { "Closed" };
+
+            void addToContent(const tgui::Widget::Ptr &item, const tgui::String &widgetName = "");
+            bool removeFromContent(const Widget::Ptr &widget);
+            const std::vector<tgui::Widget::Ptr> &getContentWidgets() const { return this->content->getWidgets(); }
```
Same thing here : How about exposing the whole `content` object through const+non-const `getContent` functions ? It even gives access to the `content` object, not just `content->widgets`. We could also just expose the `getContentWidgets` if we dont want to expose the content object

The caller can then (if non-const) use vector functions like `push_back` (or even `emplace_back` with the performance benefits associated), and if const can call `size()` or iterate normally

leo7418 replied:
Same

f-michaut replied:
So actually, after re-reviewing the code, this is fine since this is not just a wrapper around the underlying methods : there is actually extra logic involved here, so abstracting it to prevent bad usage makes total sense.

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+            void setFoldable(bool enable) { this->foldable = enable; this->setFoldImageVisible(enable); }
+            bool isFoldable() const { return foldable; }
+
+            void setFoldImageVisible(bool visible) { this->icon->setVisible(visible); }
+            bool isFoldImageVisible() const { return this->icon->isVisible(); }
```
same

leo7418 replied:
Same

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -0,0 +1,86 @@
+            void setFoldImageVisible(bool visible) { this->icon->setVisible(visible); }
+            bool isFoldImageVisible() const { return this->icon->isVisible(); }
+
+            void open(bool useAnim = false) { this->isOpened = true; this->toggleContent(useAnim); }
+            void close(bool useAnim = false) { this->isOpened = false; this->toggleContent(useAnim); }
```
Can theses 2 call a private function `do_open_close(bool open, bool useAnim)` to keep code centralized ?
It can even be declared `inline` and defined in the Header to allow compiler to optimize the function away

f-michaut commented on include/Components/InputFileDialog.hpp:
```diff
@@ -0,0 +1,44 @@
+            ~InputFileDialog();
+
+            typedef std::shared_ptr<InputFileDialog> Ptr;
+			typedef std::shared_ptr<const InputFileDialog> ConstPtr;
+
+			static InputFileDialog::Ptr create() { return std::make_shared<InputFileDialog>(); }
+			static InputFileDialog::Ptr copy(InputFileDialog::ConstPtr widget) { return widget ? std::static_pointer_cast<InputFileDialog>(widget->clone()) : nullptr; }
+
+            void setMode(FileMode mode) { this->mode = mode; }
+            FileMode getMode() const { return this->mode; }
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<InputFileDialog>(*this); }
```
indentation

f-michaut commented on include/Components/InputFileDialog.hpp:
```diff
@@ -0,0 +1,44 @@
+            InputFileDialog(const char* typeName = "Components::InputFileDialog", bool initRenderer = true);
+            ~InputFileDialog();
+
+            typedef std::shared_ptr<InputFileDialog> Ptr;
+			typedef std::shared_ptr<const InputFileDialog> ConstPtr;
+
+			static InputFileDialog::Ptr create() { return std::make_shared<InputFileDialog>(); }
+			static InputFileDialog::Ptr copy(InputFileDialog::ConstPtr widget) { return widget ? std::static_pointer_cast<InputFileDialog>(widget->clone()) : nullptr; }
+
+            void setMode(FileMode mode) { this->mode = mode; }
+            FileMode getMode() const { return this->mode; }
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<InputFileDialog>(*this); }
```
Theses `typedef` and functions (ignoring the set/get mode) are repeated across all components, is it possible to use CRTP to define them in a template class that we can just inherit from and have them without code duplication ?

leo7418 replied:
These typedefs must be present in each custom widgets

https://tgui.eu/tutorials/latest-stable/custom-widgets/#basic-functionality

f-michaut replied:
I understand they need to be in each class, but the point of CRTP would be to define them in a template class that each class would inherit from, which would include the typedefs + functions in each class, and since it's template (with CRTP) it can resolve the proper types to the child class' type

f-michaut reviewed (commented):
I haven't reviewed everything yet, cause I think theses comments will apply to most files. 

I will wait for theses comments resolution to look at the `cpp` files themselves, for now i'll focus on the structure before looking into implementation as implementation will likely change from structure changes

Sorry for the amount of comments 😅

leo7418 committed:
- Add .editorconfig
- Fix indentation
- On Foldout:
    - New function openClose
    - open and close function use now openClose

f-michaut commented on include/Components/List.hpp:
```diff
@@ -0,0 +1,51 @@
+			typedef std::shared_ptr<const List> ConstPtr;
+
+			static List::Ptr create() { return std::make_shared<List>(); }
+			static List::Ptr copy(List::ConstPtr widget) { return widget ? std::static_pointer_cast<List>(widget->clone()) : nullptr; }
```
Indentation :p

f-michaut commented on include/Components/List.hpp:
```diff
@@ -0,0 +1,51 @@
+            void addSeparator(const tgui::Layout2d &size, tgui::Color color = tgui::Color(0, 0, 0, 50));
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<List>(*this); }
```
indentation

f-michaut commented on include/Components/ListMenu.hpp:
```diff
@@ -0,0 +1,69 @@
+
+			static ListMenu::Ptr create() { return std::make_shared<ListMenu>(); }
+			static ListMenu::Ptr copy(ListMenu::ConstPtr widget) { return widget ? std::static_pointer_cast<ListMenu>(widget->clone()) : nullptr; }
+
```
T_T

f-michaut commented on include/Components/ListMenu.hpp:
```diff
@@ -0,0 +1,69 @@
+            tgui::Widget::Ptr addTitle(const tgui::String &title, const tgui::String &widgetName = "");
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<ListMenu>(*this); }
```
indentation

f-michaut commented on include/Settings/View.hpp:
```diff
@@ -0,0 +1,55 @@
+			typedef std::shared_ptr<const View> ConstPtr;
+
+			static View::Ptr create() { return std::make_shared<View>(); }
+			static View::Ptr copy(View::ConstPtr widget) { return widget ? std::static_pointer_cast<View>(widget->clone()) : nullptr; }
```
indentation

f-michaut commented on include/Settings/View.hpp:
```diff
@@ -0,0 +1,55 @@
+            const tgui::Widget::Ptr &getCurrentMenuContent() const { return this->currentMenuContent; }
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<View>(*this); }
```
indent

f-michaut commented on source/Application.cpp:
```diff
@@ -22,58 +23,60 @@ namespace FileShare::GUI {
+            if (showSettings) {
+                sider->remove(deviceListView);
+                sider->add(settingsView);
+                content->add(settingsView->getCurrentMenuContent());
             } else {
-                sider->setVisible(true);
-                sider->moveWithAnimation({ "0", "0" }, sf::milliseconds(ANIMATION_DURATION));
-                button->resizeWithAnimation({ "20%", button->getSize().y }, sf::milliseconds(ANIMATION_DURATION));
+                sider->remove(settingsView);
+                sider->add(deviceListView);
             }
```
We don't need to do `content->add(deviceListView->getCurrentMenuContent());` or similar in the `else` clause ?

leo7418 replied:
I haven't implemented this part yet, but yes, we will need it later when we display the files and folders of the device the user will be connected to

f-michaut commented on source/Application.cpp:
```diff
@@ -22,58 +23,60 @@ namespace FileShare::GUI {
-                });
+        auto settingsView = this->settingsController->getView();
+        settingsView->setAutoLayout(tgui::AutoLayout::Top);
+        settingsView->onMenuChange.connect([=](tgui::Widget::Ptr widget) {
+            content->removeAllWidgets();
+            content->add(widget);
+        });
```
Isn't this made redundant with the `button->onClick` block ?
Or maybe part of the `button->onClick` is redundant, as this should handle it instead ?

leo7418 replied:
Are you talking about the button in this same file? If so, it has nothing to do with the settings menu: it just switches between the device view and the settings

f-michaut commented on source/Application.cpp:
```diff
@@ -22,58 +23,60 @@ namespace FileShare::GUI {
+    // TODO:
+    // - File/folder picker
+    // - Création d'un dossier virtuel
+    // - En haut: chemin complet du dossier actuel -> au click, revenir en arrière
```
Remove TODO if done

f-michaut commented on source/Components/List.cpp:
```diff
@@ -0,0 +1,131 @@
+    void List::updateItemsPosition(unsigned int from)
+    {
+        auto items = this->getWidgets();
+        if (items.size() == 0) {
```
```suggestion
        if (from >= items.size()) {
```
Avoid crashes if misused. Also handle the `== 0` case, since if size == 0, then from cannot be smaller

f-michaut commented on source/Debug.cpp:
```diff
@@ -0,0 +1,42 @@
+    void Debug::debug(const tgui::Vector2<T> &vector, const tgui::String &prefix)
+    {
+        std::cout << Debug::getLog(vector, prefix) << std::endl;
+    }
```
You can't declare a non-specialized template function body in a cpp file. It needs to be declared in the hpp.

Its a C++ limitation, because of the way the compiler handles translation units. IMO it could be delayed to the linker phase, but that's not currently the case and attempting to use this function in any other file will trigger a compilation error.

It's not apparently in use, so I think it can be deleted, or needs to be fixed if it is for future use.

leo7418 replied:
I won't delete it because I need it for debugging since TGui hasn't developed the debug part of the classes much.

However, it's strange because I used it during my development and didn't get any compilation errors.
But in any case, I'll put it in the .hpp

f-michaut commented on source/DeviceList/View.cpp:
```diff
@@ -86,56 +86,16 @@ namespace FileShare::GUI::DeviceList {
-        layout->add(labelTitle);
-
+        this->addTitle(title);
         for (ushort i = 0; i < options.size(); i += 1) {
```
Use a range-based for loop instead : `for(auto option : options)`

f-michaut commented on source/Components/Button.cpp:
```diff
@@ -0,0 +1,328 @@
+        }
+
+        if (active) {
+            this->activeSignalId = automatic ? this->onClick(&Button::toggleActive, this) : 1;
+        } else {
+            if (this->activeSignalId > 1) {
```
What does the signalId 1 do ?

leo7418 replied:
Currently, it does nothing: I just use it to have a valid value in activeSignalId, and I use it to know if there was a previously connected signal or not (1 == no signal connected)

f-michaut replied:
But can't the `onClick` return signal_id `1` ?
I don't see any guarantees in the documentation about the range of ids, won't this misbehave if tgui sometimes returns `1` as id ?

leo7418 replied:
The ID is incremented with each new signal, and TGui uses some signals in its internal functioning. So it will never return an ID = 1

f-michaut commented on source/Components/Button.cpp:
```diff
@@ -0,0 +1,328 @@
+
+    void Button::updateSize()
+    {
+        if (this->getImageScaling() >= 0) {
```
What is `getImageScaling` mean ?
I don't understand why we `return` and skip the whole function logic

leo7418 replied:
The Button class was really complicated because it uses a component system that it is the only one using (since it's in beta).
By default, the image size of a Button in TGui is relative to the Button size. But I wanted to be able to assign a fixed size to the image (among other things). So I took a lot of functions from its parent to adapt its content: `getImageScaling >= 0` allows me to know if a size relative to the Button size is applied or if it's a fixed size. If it's a relative size, then I return by calling the parent function to let it take control.

f-michaut commented on source/Components/Button.cpp:
```diff
@@ -0,0 +1,328 @@
+
+            const tgui::Outline &borders = this->m_backgroundComponent->getBorders();
+
+            Widget::setSize({
+                size.x,
+                std::round(this->m_textComponent->getLineHeight() * 1.25f) + borders.getTop() + borders.getBottom()
+            });
+
+            this->recalculateGlyphSize();
+
+            if (this->m_string.empty())
+            {
+                const tgui::Vector2f innerSize = this->m_backgroundComponent->getClientSize();
+
+                Widget::setSize({
+                    this->imageComponent->getSize().x + (innerSize.y - this->imageComponent->getSize().y) + borders.getLeft() + borders.getRight(),
+                    size.y
+                });
+            } else {
+                const float spaceAroundImageAndText = this->m_textComponent->getLineHeight();
+
+                Widget::setSize({
+                    this->imageComponent->getSize().x + this->distanceBetweenTextAndImage + this->m_textComponent->getSize().x + spaceAroundImageAndText + borders.getLeft() + borders.getRight(),
+                    size.y
+                });
```
Instead of 2 calls to `setSize`, could we compute the `size.x` and `size.y`, store them in variables and call `setSize` once at the end ?

leo7418 replied:
As mentioned above: the Button class was really complicated, so I preferred to keep the code as similar as possible to the base function code, in order to be able to modify these functions if an update to the base class code is released and is important to apply

leo7418 committed:
- Remove some deprecated code after the last release of TGui
- Fix indentation
- Remove TODO
- On Debug:
    - Move template function from cpp to hpp
- On List component:
    - updateItemsPosition return if from is superior or equal to the size of items for a double protection
- On DeviceList view:
    - createSection use now auto loop instead of a regular one

f-michaut commented on source/Components/Button.cpp:
```diff
@@ -0,0 +1,328 @@
+/*
+** Project FileShare, 2022
+**
+** Author Léo Lhuile
+**
+** Started on  Sun Feb 13 22:03:32 2022 Léo Lhuile
+** Last update Mon Aug 29 20:45:51 2022 Léo Lhuile
+**
+** Button.cpp : Implementation of Button class
+*/
```
Just realized all the headers are outdated 🙈 

There's an extension for that : https://marketplace.visualstudio.com/items?itemName=doi.fileheadercomment

leo7418 committed:
Update header informations of each files

f-michaut commented:
Just need to fix the conflicts (use rebase - not merge), and it's good to go !

leo7418 committed:
- Update tgui version to 1.4.1
